pontoon/base/models.py

```diff
@@ -43,96 +43,100 @@
-
     def with_translation_counts(self, start_date=None, query_filters=None, limit=100):
         """
-        Returns contributors list, sorted by count of their translations.
-        Every user instance has added following properties:
+        Returns contributors list, sorted by count of their translations. Every user instance has
+        the following properties:
         * translations_count
         * translations_approved_count
         * translations_unapproved_count
         * translations_needs_work_count
-        Method has been created mainly to improve performance and to optimize
-        count of sql queries during generation of metrics.
+        * user_role
+
         All counts will be returned from start_date to now().
         :param date start_date: start date for translations.
         :param django.db.models.Q query_filters: filters contributors by given query_filters.
         :param int limit: limit results to this number.
         """
-        def translations_count(query=None):
-            """Short helper to avoid duplication of passing dates."""
-            query = query or Q()
-            if start_date:
-                query &= Q(translation__date__gte=start_date)
 
-            if query_filters:
-                query &= query_filters
+        # Collect data for faster user stats calculation.
+        user_stats = {}
+        translations = Translation.objects.exclude(user=None)
 
-            return self._changed_translations_count(query)
+        if start_date:
+            translations = translations.filter(date__gte=start_date)
+        if query_filters:
+            translations = translations.filter(query_filters)
 
-        def with_roles(self):
-            """Add user roles."""
-            managers = defaultdict(set)
-            translators = defaultdict(set)
+        translations = (
+            translations
+            .values('user', 'approved', 'fuzzy')
+            .annotate(count=Count('user'))
+        )
 
-            locales = Locale.objects.prefetch_related(
-                Prefetch(
-                    'managers_group__user_set',
-                    to_attr='fetched_managers'
-                ),
-                Prefetch(
-                    'translators_group__user_set',
-                    to_attr='fetched_translators'
-                )
-            )
+        for translation in translations:
+            count = translation['count']
+            user = translation['user']
 
-            for locale in locales:
-                for user in locale.managers_group.fetched_managers:
-                    managers[user].add(locale.code)
-                for user in locale.translators_group.fetched_translators:
-                    translators[user].add(locale.code)
+            if translation['approved']:
+                status = 'approved'
+            elif translation['fuzzy']:
+                status = 'fuzzy'
+            else:
+                status = 'suggested'
```

Comment:
What about rejected translations?

Reply:
We currently don't show them on the dashboard and since this bug is only about optimization, I didn't bother adding rejected. The number of rejected translations has negative connotation, so we should be careful about displaying it publicly and communicate ir properly.

Reply:
Mmm, the thing that bothers me is that you consider rejected translations to be suggested with this code. Isn't that a bit of a problem?

Reply:
The code in production does that, too. The same thing happens in team, project, localization dashboards, and also filters: Suggested includes rejected. See [bug 1377969](https://bugzilla.mozilla.org/show_bug.cgi?id=1377969).

Reply:
Fair enough then! :)